lib/deep_cover/tools/memoize.rb

```diff
@@ -0,0 +1,85 @@
+module DeepCover
+  module Memoized
+    def self.included(base)
+      base.extend ClassMethods
+      base._setup_memoizer
+      base._existing_subclasses.each(&:_setup_memoizer)
+    end
+
+    def freeze
+      self.class.memoized.each do |method|
+        send method
+      end
+      super
+    end
+
+    module ClassMethods
+      def memoized
+        @memoized ||= [].freeze
+      end
+
+      def inherited(base)
+        base._setup_memoizer
+        base._add_memoizers(*memoized)
+        super
+      end
+
+      def memoize(*methods)
+        methods = memoizable_methods if methods.empty?
+        _add_memoizers(*methods)
+        _existing_subclasses.each{|k| k._add_memoizers(*methods) }
+      end
+
+      def memoizable_methods
+        (instance_methods - Object.instance_methods)
+        .reject{|m| instance_method(m).arity > 0}
+      end
+
+      def _setup_memoizer
+        @_memoizer = Module.new
+        const_set :Memoizer, @_memoizer
+        prepend @_memoizer
+      end
+
+      def _add_memoizers(*methods)
+        new_methods = methods - memoized
+        puts "new methods for #{self}: #{new_methods}" unless new_methods.empty?
+        methods |= memoized
+        @memoized = methods.freeze
+        _define_memoizers(*new_methods)
+      end
+
+      def _existing_subclasses
+        ObjectSpace.each_object(Class).select { |klass| klass < self }
```

Comment:
This might be an issue for jruby support... Makes it slightly more complaticated for users.

https://github.com/jruby/jruby/wiki/PerformanceTuning#dont-enable-objectspace

This is just so that you can define memoizers on a class after it has been inherited? Not quite this is of any use really...

Reply:
Mmm, good point. I'll figure something out then.